django/contrib/admin/options.py

```diff
@@ -1201,12 +1202,49 @@ def construct_search(field_name):
                     if hasattr(field, "path_infos"):
                         # Update opts to follow the relation.
                         opts = field.path_infos[-1].to_opts
+            # Return the field name as-is if it already includes __exact
+            if field_name.endswith("__exact"):
+                return field_name
             # Otherwise, use the field with icontains.
             return "%s__icontains" % field_name
 
         may_have_duplicates = False
         search_fields = self.get_search_fields(request)
         if search_fields and search_term:
+            # Add string cast annotations for non-string exact lookups
+            str_annotations = {}
+            for field in search_fields:
+                if field.endswith("__exact"):
+                    field_name = field.rsplit("__exact", 1)[0]
+                    try:
+                        field_obj = queryset.model._meta.get_field(field_name)
+                        if not isinstance(
+                            field_obj, (models.CharField, models.TextField)
+                        ):
+                            str_annotations[f"{field_name}_str"] = Cast(
+                                field_name, output_field=models.CharField()
+                            )
+                    except FieldDoesNotExist:
+                        continue
+
+            if str_annotations:
+                queryset = queryset.annotate(**str_annotations)
+                # Modify search fields to use string casts for non-string exact fields
+                search_fields = [
+                    (
+                        f"{field.rsplit('__exact', 1)[0]}_str__exact"
+                        if field.endswith("__exact")
+                        and not isinstance(
+                            queryset.model._meta.get_field(
+                                field.rsplit("__exact", 1)[0]
+                            ),
+                            (models.CharField, models.TextField),
+                        )
+                        else field
+                    )
+                    for field in search_fields
+                ]
+
             orm_lookups = [
                 construct_search(str(search_field)) for search_field in search_fields
             ]
```

Comment:
We should be able to build `orm_lookups` as we loop through the `search_fields` to avoid multiple loops through  `search_fields`
```suggestion
            # Add string cast annotations for non-string exact lookups.
            str_annotations = {}
            orm_lookups = []
            for field in search_fields:
                if field.endswith("__exact"):
                    field_name = field.rsplit("__exact", 1)[0]
                    try:
                        field_obj = queryset.model._meta.get_field(field_name)
                    except FieldDoesNotExist:
                        lookup = construct_search(field)
                        orm_lookups.append(lookup)
                        continue
                    if not isinstance(field_obj, (models.CharField, models.TextField)):
                        str_annotations[f"{field_name}_str"] = Cast(
                            field_name, output_field=models.CharField()
                        )
                        lookup = construct_search(f"{field_name}_str__exact")
                        orm_lookups.append(lookup)
                    else:
                        lookup = construct_search(field)
                        orm_lookups.append(lookup)
                else:
                    lookup = construct_search(str(field))
                    orm_lookups.append(lookup)
```

Reply:
Nice!  
Slightly more complex logic within the loop, but definitely reduces the number of iterations
let me just quickly run it locally?

Reply:
Yeah sure, feel free to simplify it also 👍